eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala

```diff
@@ -0,0 +1,276 @@
+import scodec.bits.ByteVector
+
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.reflect.ClassTag
+
+/**
+ * Child actor of a Peer that handles accepting or rejecting a channel open request initiated by a remote peer and
+ * configuring local parameters for all open channel requests. It only handles one channel request at a time.
+ * If a concurrent request comes while still evaluating a previous one, the later request is immediately rejected.
+ *
+ * Note: If the remote peer disconnects before the interceptor fails or continues the non-initiator flow, according to the
+ * Lightning spec the flow should be canceled. Therefore any response sent by this actor with a different `peerConnection`
+ * should be ignored and not forwarded to the remote peer.
+ */
+object OpenChannelInterceptor {
+
+  // @formatter:off
+  sealed trait Command
+
+  sealed trait WaitForRequestCommands extends Command
+  case class OpenChannelNonInitiator(remoteNodeId: PublicKey, open: Either[protocol.OpenChannel, protocol.OpenDualFundedChannel], localFeatures: Features[InitFeature], remoteFeatures: Features[InitFeature], peerConnection: ActorRef[Any]) extends WaitForRequestCommands {
+    val temporaryChannelId: ByteVector32 = open.fold(_.temporaryChannelId, _.temporaryChannelId)
+    val fundingAmount: Satoshi = open.fold(_.fundingSatoshis, _.fundingAmount)
+    val channelFlags: ChannelFlags = open.fold(_.channelFlags, _.channelFlags)
+    val channelType_opt: Option[ChannelType] = open.fold(_.channelType_opt, _.channelType_opt)
+  }
+  case class OpenChannelInitiator(replyTo: ActorRef[Any], remoteNodeId: PublicKey, open: Peer.OpenChannel, localFeatures: Features[InitFeature], remoteFeatures: Features[InitFeature]) extends WaitForRequestCommands
+
+  private sealed trait CheckRateLimitsCommands extends Command
+  private case class PendingChannelsRateLimiterResponse(response: PendingChannelsRateLimiter.Response) extends CheckRateLimitsCommands
+
+  private sealed trait QueryPluginCommands extends Command
+  private case class PluginOpenChannelResponse(pluginResponse: InterceptOpenChannelResponse) extends QueryPluginCommands
+  private case object PluginTimeout extends QueryPluginCommands
+
+  private sealed trait ErrorResponse
+  private object RateLimitError extends ErrorResponse {
+    override def toString: String = "rate limit reached"
+  }
+  private case class ChannelExceptionError(ex: ChannelException) extends ErrorResponse {
+    override def toString: String = ex.getMessage
+  }
+
+  private object ConcurrentRequestError extends ErrorResponse {
+    override def toString: String = "concurrent request rejected"
+  }
+  private object PluginTimeoutError extends ErrorResponse {
+    override def toString: String = "plugin timeout"
+  }
+  private case class PluginRejectedError(error: Error) extends ErrorResponse {
+    override def toString: String = error.toAscii
+  }
```

Comment:
Each of these errors is used exactly once, and only used for its `toString` value. This is an indication that it's probably overly verbose to define this trait, I would remove them completely and just change the signature of `sendFailure` to:

```
private def sendFailure(error: String, nonInitiator: OpenChannelNonInitiator): Unit
```

In each error case, you would just inline the error message.

Reply:
fixed in [3c4137f](https://github.com/ACINQ/eclair/pull/2552/commits/3c4137fdb0bb6c8f712c40f508c9fc0468c13304).